File: app/src/main/java/com/example/tryanderror/CategoryAdapter.java
package com.example.tryanderror;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;

public class CategoryAdapter extends RecyclerView.Adapter<CategoryAdapter.ViewHolder> {

    public interface OnCategoryClickListener {
        void onCategoryClick(Category category);
    }

    private List<Category> categories;
    private OnCategoryClickListener listener;
    private int selectedPosition = 0;

    public CategoryAdapter(List<Category> categories, OnCategoryClickListener listener) {
        this.categories = categories;
        this.listener = listener;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        CardView card;
        TextView tvName;

        public ViewHolder(View view) {
            super(view);
            card = view.findViewById(R.id.categoryCard);
            tvName = view.findViewById(R.id.tvCategoryName);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_category, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Category category = categories.get(position);
        holder.tvName.setText(category.getEmoji() + "  " + category.getName());

        if (position == selectedPosition) {
            // Selected: vibrant brand color
            holder.card.setCardBackgroundColor(Color.parseColor(category.getColorHex()));
            holder.tvName.setTextColor(ContextCompat.getColor(holder.itemView.getContext(), android.R.color.white));
        } else {
            // Unselected: adapts to dark/light mode via color resource
            holder.card.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.category_unselected_bg));
            holder.tvName.setTextColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.category_unselected_text));
        }

        holder.itemView.setOnClickListener(v -> {
            int prev = selectedPosition;
            selectedPosition = holder.getAdapterPosition();
            notifyItemChanged(prev);
            notifyItemChanged(selectedPosition);
            if (listener != null) {
                listener.onCategoryClick(category);
            }
        });
    }

    @Override
    public int getItemCount() {
        return categories != null ? categories.size() : 0;
    }

    public void selectCategory(String key) {
        int idx = -1;
        for (int i = 0; i < categories.size(); i++) {
            if (categories.get(i).getKey().equals(key)) {
                idx = i;
                break;
            }
        }
        if (idx != -1) {
            int prev = selectedPosition;
            selectedPosition = idx;
            notifyItemChanged(prev);
            notifyItemChanged(selectedPosition);
        }
    }
}

File: app/src/main/java/com/example/tryanderror/CategoryAdapter.kt
package com.example.tryanderror

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView

data class Category(val name: String, val key: String, val emoji: String, val colorHex: String)

class CategoryAdapter(
    private val categories: List<Category>,
    private val onCategoryClick: (Category) -> Unit
) : RecyclerView.Adapter<CategoryAdapter.ViewHolder>() {

    private var selectedPosition = 0

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val card: CardView = view.findViewById(R.id.categoryCard)
        val tvName: TextView = view.findViewById(R.id.tvCategoryName)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_category, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val category = categories[position]
        holder.tvName.text = "${category.emoji}  ${category.name}"

        if (position == selectedPosition) {
            // Selected: vibrant brand color
            holder.card.setCardBackgroundColor(android.graphics.Color.parseColor(category.colorHex))
            holder.tvName.setTextColor(androidx.core.content.ContextCompat.getColor(holder.itemView.context, android.R.color.white))
        } else {
            // Unselected: adapts to dark/light mode via color resource
            holder.card.setCardBackgroundColor(androidx.core.content.ContextCompat.getColor(holder.itemView.context, R.color.category_unselected_bg))
            holder.tvName.setTextColor(androidx.core.content.ContextCompat.getColor(holder.itemView.context, R.color.category_unselected_text))
        }

        holder.itemView.setOnClickListener {
            val prev = selectedPosition
            selectedPosition = holder.adapterPosition
            notifyItemChanged(prev)
            notifyItemChanged(selectedPosition)
            onCategoryClick(category)
        }
    }

    override fun getItemCount() = categories.size

    fun selectCategory(key: String) {
        val idx = categories.indexOfFirst { it.key == key }
        if (idx != -1) {
            val prev = selectedPosition
            selectedPosition = idx
            notifyItemChanged(prev)
            notifyItemChanged(selectedPosition)
        }
    }
}
